hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -727,6 +702,374 @@ func assignOptionalEnumTypeFromOptionalEnumType(
+// assignOptionalObjectTypeFromObjectType will generate a conversion if both properties are
+// TypeNames referencing ObjectType definitions and only the destination is optional
+//
+// For ConvertFrom:
+//
+// var <local> <destinationType>
+// <local>.ConvertFrom(<source>)
+// <destination> = &<local>
+//
+// For ConvertTo:
+//
+// var <local> <destinationType>
+// <source>.ConvertTo(&<local>)
+// <destination> = &<local>
+//
+func assignOptionalObjectTypeFromObjectType(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint,
+	conversionContext *StorageConversionContext) StorageTypeConversion {
+
+	// Require source to be non-optional
+	if _, srcOpt := AsOptionalType(sourceEndpoint.Type()); srcOpt {
+		return nil
+	}
+
+	// Require destination to be optional
+	if _, dstOpt := AsOptionalType(destinationEndpoint.Type()); !dstOpt {
+		return nil
+	}
+
+	// Require source to be an object
+	if _, srcIsObject := conversionContext.ResolveObject(sourceEndpoint.Type()); !srcIsObject {
+		return nil
+	}
+
+	// Require destination to be an object
+	dstName, dstIsObject := conversionContext.ResolveObject(destinationEndpoint.Type())
+	if !dstIsObject {
+		return nil
+	}
+
+	copyvar := destinationEndpoint.CreateLocal()
+
+	return func(reader dst.Expr, writer dst.Expr, generationContext *CodeGenerationContext) []dst.Stmt {
+
+		localId := dst.NewIdent(copyvar)
+		errLocal := dst.NewIdent("err")
+
+		declaration := astbuilder.LocalVariableDeclaration(copyvar, createTypeDeclaration(dstName, generationContext), "")
+
+		var conversion dst.Stmt
+		if dstName.PackageReference.Equals(generationContext.CurrentPackage()) {
+			conversion = astbuilder.SimpleAssignment(
+				errLocal,
+				token.DEFINE,
+				astbuilder.CallExpr(localId, conversionContext.functionName, astbuilder.AddrOf(reader)))
+		} else {
+			conversion = astbuilder.SimpleAssignment(
+				errLocal,
+				token.DEFINE,
+				astbuilder.CallExpr(reader, conversionContext.functionName, astbuilder.AddrOf(localId)))
+		}
+
+		checkForError := astbuilder.ReturnIfNotNil(
+			errLocal,
+			astbuilder.WrappedErrorf(
+				"populating %s from %s, calling %s()",
+				destinationEndpoint.name, sourceEndpoint.name, conversionContext.functionName))
+
+		assignment := astbuilder.SimpleAssignment(
+			writer,
+			token.ASSIGN,
+			astbuilder.AddrOf(dst.NewIdent(copyvar)))
+
+		return []dst.Stmt{
+			declaration,
+			conversion,
+			checkForError,
+			assignment,
+		}
+	}
+}
+
+// assignOptionalObjectTypeFromOptionalObjectType will generate a conversion if both properties are
+// TypeNames referencing ObjectType definitions and both source and destination are optional
+//
+// For ConvertFrom:
+//
+// var <local> <destinationType>
+// if <source> != nil {
+//     <local>.ConvertFrom(<source>)
+// }
+// <destination> = &<local>
+//
+// For ConvertTo:
+//
+// var <local> <destinationType>
+// if <source> != nil {
+//     <source>.ConvertTo(&<local>)
+// }
+// <destination> = &<local>
```

Comment:
Shouldn't nil be assigned to destination if source is nil? This always sets destination.

```suggestion
// 
// if <source> != nil {
//     var <local> <destinationType>
//    <local>.ConvertFrom(<source>)
//    <destination> = &<local>
// }
//
// For ConvertTo:
//
// if <source> != nil {
//     var <local> <destinationType>
//     <source>.ConvertTo(&<local>)
//     <destination> = &<local>
// }
```

Reply:
(Again, the golden files are right for this, it's just that the comment is misleading.)